File: bindings/lua/loop/worker.lua
--- The worker: its body (one warmup iteration, the warmup barrier,
--- the main loop), one iteration, the session pump loop the stream
--- shape drives, and the round-trip comparison that decides between a
--- worker error and a data mismatch.

local ops = require "ops"
local payload = require "payload"
local size = require "size"
local state = require "state"

local M = {}

-- Pump loop. The Go harness hands ITB an io.Reader / io.Writer pair
-- and ITB drives the chunk loop internally; the C ABI has no reader /
-- writer entry, so the caller drives it: open a session, feed slices of
-- at most 1 MiB, drain whatever the session has produced after every
-- write (a read before end never blocks), end, then drain until the
-- session reports finished (after end, a read on an empty spool blocks
-- until the terminal bytes arrive). The loop is written here rather
-- than delegated to the binding's pump convenience so it stands in the
-- utility, at the same place, in every language.
local function pump(pipe, encrypt, src)
    local session = encrypt and pipe:encrypt_stream() or pipe:decrypt_stream()
    local ok, res = pcall(function()
        local parts = {}
        local off = 1
        local total = #src
        while off <= total do
            local last = math.min(off + state.PUMP_SLICE - 1, total)
            session:write(src:sub(off, last))
            off = last + 1
            while true do
                local chunk = session:read(state.PUMP_SLICE)
                if #chunk == 0 then
                    break
                end
                parts[#parts + 1] = chunk
            end
        end
        session:finish()
        while true do
            local chunk, finished = session:read(state.PUMP_SLICE)
            if #chunk > 0 then
                parts[#parts + 1] = chunk
            end
            if finished then
                break
            end
        end
        return table.concat(parts)
    end)
    session:free()
    if not ok then
        error(res, 0)
    end
    return res
end

-- First offset at which a and b differ; the shorter length when one is
-- a prefix of the other.
local function first_difference(a, b)
    local n = math.min(#a, #b)
    for i = 1, n do
        if a:byte(i) ~= b:byte(i) then
            return i - 1
        end
    end
    return n
end

-- Up to 16 bytes of buf from off (zero-based) as lowercase hex, or "-"
-- when buf has no bytes there.
local function hex_window(buf, off)
    if off >= #buf then
        return "-"
    end
    local out = {}
    for i = off + 1, math.min(off + 16, #buf) do
        out[#out + 1] = string.format("%02x", buf:byte(i))
    end
    return table.concat(out)
end

-- Records a worker error for a failed cipher call.
local function cipher_fail(w, it, shape, direction, err)
    state.worker_fail(w, string.format("g%d iter %d shape=%s: %s: %s",
        w.id, it, state.shape_name(shape), direction, state.status_detail(err)))
end

--- One iteration. In order: refill the plaintext under rotating mode;
--- take the read lock; pick the surface; encrypt (timed); decrypt
--- (timed); compare the round-trip with the plaintext; bump the
--- counters; release the lock. The whole round-trip is kept clear of
--- handle-mutating maintenance (rekey, blob reopen), which runs after
--- this returns, from the worker loop, so nothing lands between an
--- encrypt and its matching decrypt. On this binding that separation
--- needs no lock: the run is one thread, so the ordering is the only
--- mechanism there is. Returns false after recording the worker error.
function M.iterate(w, it)
    local r = w.run

    if w.payload_mode == payload.ROTATING then
        w.plaintext, w.rng =
            payload.fill(payload.ROTATING, w.seeded, w.rng, #w.plaintext)
    end

    -- Shape dispatch. message is one whole-buffer call on the Single
    -- Message Pipeline; stream_one_shot is one whole-buffer call on the
    -- streaming Pipeline (the C ABI's ITB_Triple_EncryptStream, which
    -- routes to the same whole-buffer stream entry the Go harness calls
    -- by name); stream opens a session on the same streaming Pipeline
    -- and drives the chunk loop from here. Under both the three rotate
    -- by iteration number so the session path and the whole-buffer path
    -- alternate on one handle inside every worker — the cross-path
    -- state-reuse hazard this harness exists to catch.
    local shape = r.cfg.shape
    if shape == state.SHAPE_BOTH then
        shape = ({ state.SHAPE_STREAM, state.SHAPE_MESSAGE,
            state.SHAPE_STREAM_ONE_SHOT })[it % 3 + 1]
    end

    local want = w.plaintext
    local wire, got
    if shape == state.SHAPE_STREAM then
        local t0 = size.now_ns()
        local ok, res = pcall(pump, r.stream_pipe, true, want)
        if not ok then
            cipher_fail(w, it, shape, "encrypt", res)
            return false
        end
        wire = res
        w.nanos_enc = w.nanos_enc + (size.now_ns() - t0)
        t0 = size.now_ns()
        ok, res = pcall(pump, r.stream_pipe, false, wire)
        if not ok then
            cipher_fail(w, it, shape, "decrypt", res)
            return false
        end
        got = res
        w.nanos_dec = w.nanos_dec + (size.now_ns() - t0)
    else
        local pipe, enc, dec
        if shape == state.SHAPE_MESSAGE then
            pipe = r.msg_pipe
            enc, dec = pipe.encrypt_message, pipe.decrypt_message
        else
            pipe = r.stream_pipe
            enc, dec = pipe.encrypt_stream_one_shot, pipe.decrypt_stream_one_shot
        end
        local t0 = size.now_ns()
        local ok, res = pcall(enc, pipe, want)
        if not ok then
            cipher_fail(w, it, shape, "encrypt", res)
            return false
        end
        wire = res
        w.nanos_enc = w.nanos_enc + (size.now_ns() - t0)
        t0 = size.now_ns()
        ok, res = pcall(dec, pipe, wire)
        if not ok then
            cipher_fail(w, it, shape, "decrypt", res)
            return false
        end
        got = res
        w.nanos_dec = w.nanos_dec + (size.now_ns() - t0)
    end

    -- Failure model. A cipher call that returns a non-OK status is a
    -- worker error: it is recorded, the run is asked to stop, and the
    -- error is listed in the summary with the FAIL verdict. A
    -- round-trip that returns OK with different bytes is a data
    -- mismatch: the process terminates here, without summary or
    -- cleanup, because the Pipeline state that produced the wrong bytes
    -- is the evidence and nothing that runs afterwards may touch it.
    if got ~= want then
        local off = first_difference(want, got)
        io.stderr:write(string.format(
            "loop: DATA MISMATCH g%d iter %d shape=%s: want %d bytes, "
            .. "got %d bytes, first difference at offset %d: want %s got %s\n",
            w.id, it, state.shape_name(shape), #want, #got, off,
            hex_window(want, off), hex_window(got, off)))
        io.stderr:flush()
        -- Lua-specific. os.exit's second argument asks the interpreter
        -- to close the state on the way out, which would run every
        -- __gc metamethod, including the ones that free the Pipeline
        -- handles. Leaving it out is what "no summary, no cleanup"
        -- asks for.
        os.exit(3)
    end

    w.iters = w.iters + 1
    w.bytes_enc = w.bytes_enc + #want
    w.bytes_dec = w.bytes_dec + #got
    return true
end

--- The worker body: one warmup iteration, then the main loop until a
--- stop is requested, the duration deadline passes, or the fixed
--- per-worker iteration budget (warmup included) is spent.
---
--- Concurrency mode. This binding runs single: a stock Lua 5.4
--- interpreter has one thread and one lua_State, and its standard
--- library offers no thread primitive — coroutines are cooperative, so
--- they would run on the same interpreter one at a time and could not
--- put two calls into the library at once. --goroutines is therefore
--- accepted, clamped to 1, and reported next to the requested value,
--- so a fleet reading the summary sees the mode rather than inferring
--- it. The warmup barrier below degenerates to the single worker's own
--- first iteration.
function M.warmup(w)
    local ok, res = pcall(M.iterate, w, 0)
    if not ok then
        state.worker_fail(w, string.format("g%d iter 0: %s", w.id, tostring(res)))
        return false
    end
    return res
end

--- The main loop, entered after the warmup barrier has been passed and
--- the baselines taken.
function M.run_worker(w)
    local r = w.run
    local cfg = r.cfg
    local it = 1
    while true do
        if cfg.iterations > 0 and it >= cfg.iterations then
            break
        end
        if r.stop then
            break
        end
        if cfg.iterations == 0
            and size.now_ns() - r.start_ns >= cfg.duration_ns then
            break
        end
        local ok, res = pcall(M.iterate, w, it)
        if not ok then
            state.worker_fail(w, string.format("g%d iter %d: %s",
                w.id, it, tostring(res)))
            break
        end
        if not res then
            break
        end
        ok, res = pcall(ops.worker_maintenance, w, it)
        if not ok then
            state.worker_fail(w, string.format("g%d iter %d: %s",
                w.id, it, tostring(res)))
            break
        end
        if not res then
            break
        end
        it = it + 1
    end
end

return M

File: bindings/r/loop/worker.R
# The worker: its body (one warmup iteration, the warmup barrier, the
# main loop), one iteration, the session pump loop the stream shape
# drives, and the round-trip comparison that decides between a worker
# error and a data mismatch.

# Pump loop. The Go harness hands ITB an io.Reader / io.Writer pair and
# ITB drives the chunk loop internally; the C ABI has no reader /
# writer entry, so the caller drives it: open a session, feed slices of
# at most 1 MiB, drain whatever the session has produced after every
# write (a read before end never blocks), end, then drain until the
# session reports finished (after end, a read on an empty spool blocks
# until the terminal bytes arrive). The loop is written here rather
# than delegated to the binding's pump convenience so it stands in the
# utility, at the same place, in every language.
#
# R-specific. The slice is addressed through stream_write_slice, which
# feeds a window of the source vector without materialising it: R's
# ordinary subsetting would build a million-element index vector and a
# copy of the window for every write, which is payload-sized traffic
# the other bindings' slices do not carry.
pump <- function(pipe, encrypt, src) {
  session <- if (encrypt) {
    libitb3r::stream_encryptor(pipe)
  } else {
    libitb3r::stream_decryptor(pipe)
  }
  on.exit(libitb3r::stream_free(session), add = TRUE)
  parts <- list()
  off <- 0
  total <- length(src)
  while (off < total) {
    take <- min(PUMP_SLICE, total - off)
    libitb3r::stream_write_slice(session, src, off, take)
    off <- off + take
    repeat {
      got <- libitb3r::stream_read(session, PUMP_SLICE)
      if (length(got$chunk) == 0L) {
        break
      }
      parts[[length(parts) + 1L]] <- got$chunk
    }
  }
  libitb3r::stream_end(session)
  repeat {
    got <- libitb3r::stream_read(session, PUMP_SLICE)
    if (length(got$chunk) > 0L) {
      parts[[length(parts) + 1L]] <- got$chunk
    }
    if (isTRUE(got$finished)) {
      break
    }
  }
  if (length(parts) == 0L) {
    return(raw(0))
  }
  unlist(parts, use.names = FALSE)
}

# First offset at which a and b differ; the shorter length when one is
# a prefix of the other.
first_difference <- function(a, b) {
  n <- min(length(a), length(b))
  if (n > 0L) {
    diff <- which(a[seq_len(n)] != b[seq_len(n)])
    if (length(diff) > 0L) {
      return(diff[1L] - 1L)
    }
  }
  n
}

# Up to 16 bytes of buf from off (zero-based) as lowercase hex, or "-"
# when buf has no bytes there.
hex_window <- function(buf, off) {
  if (off >= length(buf)) {
    return("-")
  }
  last <- min(off + 16L, length(buf))
  paste(sprintf("%02x", as.integer(buf[(off + 1L):last])), collapse = "")
}

# Records a worker error for a failed cipher call.
cipher_fail <- function(w, it, shape, direction, e) {
  worker_fail(w, sprintf("g%d iter %.0f shape=%s: %s: %s",
    w$id, it, shape_name(shape), direction, status_detail(e)))
}

# One iteration. In order: refill the plaintext under rotating mode;
# take the read lock; pick the surface; encrypt (timed); decrypt
# (timed); compare the round-trip with the plaintext; bump the
# counters; release the lock. The whole round-trip is kept clear of
# handle-mutating maintenance (rekey, blob reopen), which runs after
# this returns, from the worker loop, so nothing lands between an
# encrypt and its matching decrypt. On this binding that separation
# needs no lock: the run is one thread, so the ordering is the only
# mechanism there is. Returns FALSE after recording the worker error.
iterate <- function(w, it) {
  r <- w$run

  if (w$payload_mode == PAYLOAD_ROTATING) {
    w$plaintext <- fill_payload(PAYLOAD_ROTATING, w$seeded, length(w$plaintext))
  }

  # Shape dispatch. message is one whole-buffer call on the Single
  # Message Pipeline; stream_one_shot is one whole-buffer call on the
  # streaming Pipeline (the C ABI's ITB_Triple_EncryptStream, which
  # routes to the same whole-buffer stream entry the Go harness calls
  # by name); stream opens a session on the same streaming Pipeline and
  # drives the chunk loop from here. Under both the three rotate by
  # iteration number so the session path and the whole-buffer path
  # alternate on one handle inside every worker — the cross-path
  # state-reuse hazard this harness exists to catch.
  shape <- r$cfg$shape
  if (shape == SHAPE_BOTH) {
    shape <- c(SHAPE_STREAM, SHAPE_MESSAGE, SHAPE_STREAM_ONE_SHOT)[it %% 3 + 1]
  }

  want <- w$plaintext
  if (shape == SHAPE_STREAM) {
    t0 <- now_ns()
    wire <- tryCatch(pump(r$stream_pipe, TRUE, want),
      error = function(e) {
        cipher_fail(w, it, shape, "encrypt", e)
        NULL
      })
    if (is.null(wire)) {
      return(FALSE)
    }
    w$nanos_enc <- w$nanos_enc + (now_ns() - t0)
    t0 <- now_ns()
    got <- tryCatch(pump(r$stream_pipe, FALSE, wire),
      error = function(e) {
        cipher_fail(w, it, shape, "decrypt", e)
        NULL
      })
    if (is.null(got)) {
      return(FALSE)
    }
    w$nanos_dec <- w$nanos_dec + (now_ns() - t0)
  } else {
    if (shape == SHAPE_MESSAGE) {
      pipe <- r$msg_pipe
      enc <- libitb3r::pipeline_encrypt_message
      dec <- libitb3r::pipeline_decrypt_message
    } else {
      pipe <- r$stream_pipe
      enc <- libitb3r::pipeline_encrypt_stream_one_shot
      dec <- libitb3r::pipeline_decrypt_stream_one_shot
    }
    t0 <- now_ns()
    wire <- tryCatch(enc(pipe, want),
      error = function(e) {
        cipher_fail(w, it, shape, "encrypt", e)
        NULL
      })
    if (is.null(wire)) {
      return(FALSE)
    }
    w$nanos_enc <- w$nanos_enc + (now_ns() - t0)
    t0 <- now_ns()
    got <- tryCatch(dec(pipe, wire),
      error = function(e) {
        cipher_fail(w, it, shape, "decrypt", e)
        NULL
      })
    if (is.null(got)) {
      return(FALSE)
    }
    w$nanos_dec <- w$nanos_dec + (now_ns() - t0)
  }

  # Failure model. A cipher call that returns a non-OK status is a
  # worker error: it is recorded, the run is asked to stop, and the
  # error is listed in the summary with the FAIL verdict. A round-trip
  # that returns OK with different bytes is a data mismatch: the
  # process terminates here, without summary, because the Pipeline
  # state that produced the wrong bytes is the evidence and nothing
  # that runs afterwards may touch it.
  if (!identical(got, want)) {
    off <- first_difference(want, got)
    emit(stderr(), sprintf(
      paste0("loop: DATA MISMATCH g%d iter %.0f shape=%s: want %.0f bytes, ",
        "got %.0f bytes, first difference at offset %.0f: want %s got %s\n"),
      w$id, it, shape_name(shape), length(want), length(got), off,
      hex_window(want, off), hex_window(got, off)))
    # R-specific. quit() is base R's only way out of a process, and it
    # runs the finalizers registered for exit before the process
    # leaves — which for this binding means the Pipeline handles are
    # released on the way. There is no immediate-termination entry to
    # reach for instead, so the state the mismatch produced does not
    # survive the exit here as it does in the languages that have one.
    # Nothing else runs: no summary is printed and no further cipher
    # call touches the handle.
    quit(save = "no", status = 3, runLast = FALSE)
  }

  w$iters <- w$iters + 1
  w$bytes_enc <- w$bytes_enc + length(want)
  w$bytes_dec <- w$bytes_dec + length(got)
  TRUE
}

# The worker body: one warmup iteration, then the main loop until a
# stop is requested, the duration deadline passes, or the fixed
# per-worker iteration budget (warmup included) is spent.
#
# Concurrency mode. This binding runs single: R evaluates on one thread
# and its C API is not re-entrant, so no second execution unit can be
# inside a .Call at the same time; base R carries no threading entry,
# and the parallel package answers the question by forking whole
# processes, which share no Pipeline handle. --goroutines is therefore
# accepted, clamped to 1, and reported next to the requested value, so
# a fleet reading the summary sees the mode rather than inferring it.
# The warmup barrier below degenerates to the single worker's own first
# iteration.
warmup_worker <- function(w) {
  tryCatch(iterate(w, 0),
    error = function(e) {
      worker_fail(w, sprintf("g%d iter 0: %s", w$id, conditionMessage(e)))
      FALSE
    })
}

# The main loop, entered after the warmup barrier has been passed and
# the baselines taken.
run_worker <- function(w) {
  r <- w$run
  cfg <- r$cfg
  it <- 1
  repeat {
    if (cfg$iterations > 0 && it >= cfg$iterations) {
      break
    }
    if (r$stop) {
      break
    }
    if (cfg$iterations == 0 && now_ns() - r$start_ns >= cfg$duration_ns) {
      break
    }
    # An iteration and the maintenance that follows it run with
    # interrupts suspended, so a Ctrl-C arriving inside either is held
    # until it has finished; the caller catches it at the loop's own
    # check point and turns it into the stop request.
    ok <- tryCatch(suspendInterrupts(iterate(w, it)),
      error = function(e) {
        worker_fail(w, sprintf("g%d iter %.0f: %s", w$id, it,
          conditionMessage(e)))
        FALSE
      })
    if (!isTRUE(ok)) {
      break
    }
    ok <- tryCatch(suspendInterrupts(worker_maintenance(w, it)),
      error = function(e) {
        worker_fail(w, sprintf("g%d iter %.0f: %s", w$id, it,
          conditionMessage(e)))
        FALSE
      })
    if (!isTRUE(ok)) {
      break
    }
    it <- it + 1
  }
  invisible(NULL)
}
